### ISSUE TITLE
[Android] Unable to fetch/update doc on background task (HeadlessJS)

### ISSUE THREAD

**Issue by Linoa65** [invertase/react-native-firebase#2514]
## Issue

Hi to this insane community !

I have an issue in a particular case. I use headlessJS to catch data-only message, then display notification to user. Everything works fine, but now I want to do more in this function. I need to update some docs on my firestore, after receiving a notification, and it works for foreground mode.

When I put my app in background, or closed state, the execution starts, and when it encounter a promise related to firebase things, it blocks the code and wait me to open app manually before ending the task.

A weird things is that, only promises are blocked, but the notification will be displayed (the first notification take some time, but after that it works good).

```js
export default async (message) => {
    firebaseService.onAuthStateChanged((user) => {
        if (user) {
            // Inside this method I have my logic of displaying notifications and updating database
            notificationService.onMessage(message, user._user.uid);
        }
    });
    return Promise.resolve();
};

```

Example : When I receive message, first I call my database to make an update, then I display my notification. In this case, notification will be triggered, but my update to database will never end, and if I wait too much before launching my app, the update will never be done.

I share only files needed to help, because the project is private, and I don't want to publish confidential things. If other info are needed, feel free to ask !

This problem is blocking for me, and I start to try to fix since some weeks now...

Sincerely
Flo'

---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
buildscript {
    ext {
        buildToolsVersion = "28.0.3"
        minSdkVersion = 19
        compileSdkVersion = 28
        targetSdkVersion = 28
        supportLibVersion = "28.0.0"
        supportVersion = "28.0.0"

        googlePlayServicesLocationVersion = "16.0.0"
        googlePlayServicesVersion = "16.0.0"
        googlePlayServicesVisionVersion = "17.0.2"

        firebaseVersion = "17.5.0"
        reactNativeVersion = "0.59.8"
        multiDexEnabled = true
    }
    repositories {
        google()
        maven { url "https://maven.google.com" }
        jcenter()

        maven { url "https://maven.fabric.io/public" }
        maven { url 'https://dl.bintray.com/tzutalin/maven' }
        maven { url "$rootDir/../node_modules/react-native/android" }

    }

    dependencies {
        classpath "com.android.tools.build:gradle:3.3.2"
        classpath "com.google.gms:google-services:4.2.0"
        classpath "io.fabric.tools:gradle:1.+"
        classpath "com.google.firebase:firebase-plugins:1.1.1"
        classpath "com.github.dcendents:android-maven-gradle-plugin:2.0"

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

subprojects {
    afterEvaluate { project ->
        if (project.hasProperty("android")) {
            android {
                compileSdkVersion 28
                buildToolsVersion "28.0.3"
            }
        }
    }
}


allprojects {
    repositories {
        mavenCentral()
        mavenLocal()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url "$rootDir/../node_modules/react-native/android"
        }
        maven {
            url "http://repo1.maven.org/maven2"
        }
        maven {
            url "https://jitpack.io"
        }
        maven {
            url "https://maven.google.com"
            name "Google"
        }
        google()
        jcenter()
    }
}

wrapper {
    gradleVersion = "5.4"
    distributionUrl = distributionUrl.replace("bin", "all")
}
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
  React Native Environment Info:
    System:
      OS: macOS 10.14.3
      CPU: (8) x64 Intel(R) Core(TM) i7-4750HQ CPU @ 2.00GHz
      Memory: 157.25 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.12.0 - /usr/local/bin/node
      npm: 6.9.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.1, macOS 10.14, tvOS 12.1, watchOS 5.1
      Android SDK:
        API Levels: 22, 23, 25, 26, 27, 28
        Build Tools: 27.0.2, 27.0.3, 28.0.2, 28.0.3
        System Images: android-27 | Google APIs Intel x86 Atom, android-28 | Google APIs Intel x86 Atom
    IDEs:
      Android Studio: 3.2 AI-181.5540.7.32.5056338
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.8.3 => 16.8.3 
      react-native: 0.59.8 => 0.59.8 
    npmGlobalPackages:
      create-react-native-app: 2.0.2
      react-native-cli: 2.0.1
      react-native-git-upgrade: 0.2.7
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [x] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.4.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
Interesting. Have you tried tracing the app in AndroidStudio to see where it's hanging? I see "You can do anything in your task such as network requests, timers and so on, as long as it doesn't touch UI" from https://facebook.github.io/react-native/docs/headless-js-android.html and that seems like a good assumption to test. My hypothesis is: you are touching the UI

My caveat to that hypothesis is to add ", somehow", and also possibly ", or actually some library is touching the UI on your behalf when you're not expecting it" etc

**Comment by mikehardy** [comment]
Also you did not fully fill out the template, but I wouldn't waste time trying to solve a bug when you're not current. react-native 0.59.10, react-native-firebase 5.5.6, and underlying SDKs up to date at minimum (it all works as long as you use jetifier - I test it with https://github.com/mikehardy/rnfbdemo/blob/master/make-demo-rn59.sh)

I fully understand the RN59 to RN60 upgrade hurt a lot, but as this may be a react-native issue you might want to plan getting current there in case you hit a wall trying to find the hang

**Comment by Ehesp** [comment]
That code looks suspect to me anyway. It will resolve straight away and I'm not sure what's going to happen with the listener once the headless task has resolved.

```js

export default async (message) => {
	return new Promise((resolve) => {
 		firebaseService.onAuthStateChanged((user) => {
			// do stuff...

			return resolve();
	    });
	});

};
```

Is your issue the same when the app is backgrounded and killed? Or just one?

> Hi to this insane community !

😆

**Comment by Linoa65** [comment]
I will make test and come back after with results. It's impossible for me to touch the UI, me onMessage method is only building a notification, displaying it, and playing a sound.

I'll make the change proposed by Ehesp without effect... But I'm sure, that it concerns only database promises. 

```js
        console.log("start...");
        return this.ref.doc(conversationId).get().then((snapshot) => {
            console.log(snapshot);
            console.log("end...");
        }
```

For example, this code, called from my background task is never resolved. the log "start..." will be printed and after that nothing until I start the app. I will do more test and add my results and hypothesis as soon as I can.

I will try to upgrade to RN 60 to and test what will happens.

**Comment by Linoa65** [comment]
I actually have no time to try to update to RN 0.60, but now, when I receive a notification with app closed, so the same method is called, but NO promise are resolving, they just stayed pending until something happens, but if I send a new notification, both of them are now printed, so promise are now resolved, only if one was pending just before, its not understandable... 

I'm not touching UI, I make my process inside services classes, which are not components or views...

This bug is redundant for me, so I can't correctly execute code, especially when app is on dead state, I'm unable to notify user correctly...

**Comment by Linoa65** [comment]
Trying to upgrade to RN60, but the app not build if I remove all olds link to use auto-linking, the "io.invertase.firebase.messaging" is not found, and if I try to use linking like before, I got a ClassNotFoundException on "io.invertase.firebase.messaging.RNFirebaseMessagingService"

**Comment by mikehardy** [comment]
@Linoa65 that's a separate issue I think. For RN60 even with auto-linking you still need to [depend on](https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh#L89), [import](https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh#L91), and [load](https://github.com/mikehardy/rnfbdemo/blob/master/make-demo.sh#L93) the specific sub-packages one by one. Auto-linking only handles the 'main' RNFirebasePackage. Follow those links for the necessary items. This is all [documented for messaging](https://rnfirebase.io/docs/v5.x.x/messaging/android) and each package now thanks to a recent docs PR from a fellow user

**Comment by Linoa65** [comment]
Finally upgraded on RN 0.60. After testing some modules to try to execute background task, nothing works. Same bunches of code are not executed. It concerns essentially firebase promises (even if I get logged just before trying to resole these promises) and sometimes, the displaying of notification (but I think it's linked to promises problem).

Please, tell me which infos are needed to help to find this bug or my errors. Maybe try with v6 of rn-firebase ?

**Comment by Linoa65** [comment]
After hard testing, I can affirm that only the first message is bad processed. When I receive first message after app closed, firebase promises will be unresolved until I open the app, but the nexts message ore correctly processed, my docs are update and notification displayed.

**Comment by mindtheworld** [comment]
@Linoa65 what happens when your app is closed while receiving a data-only message? would your app gets auto start and brings itself to the foreground? That's the weird behaviour I got, is this an expected behaviour? How could I move it to the background or maybe exit it gracefully? 

versions used:
 "react-native-firebase": "^5.5.6"

**Comment by Linoa65** [comment]
After more test, I try to know at which moment the code pause its execution. Whenever it encounter a Promise, the first background notification in background task is not displayed. Even a Promise.resolve() stays in pause and finished only after another data-only message is received or if I open my app.

I'm working on a solution for my case, sending remote notification in case of app closed (able to detect this with "onDisconnect" method from database module).

**Comment by Linoa65** [comment]
@mikehardy Hi ! 

I'm coming back about this issue which persists and I'm unable to fix that because I can't understand what happens really. Please tell me if I need to make again an issue with more details, but I can ensure that when I use a notification action, everything works as expected, until I make a firebase request.

Execution reached the call of promise, but get stuck until I open app. I can ensure too there is no UI updates, only get/update on database.

When I receive notification, I'm updating Firestore for some reason with no problems, promise get resolved and code execution continue until the end. The problems persists only for notification action.

I really need help on this problem, maybe v6 can fix that, but I need to wait notification/messaging package fully integrated (like it's said, it will take 2-3 weeks).

Thanks
Florent
